Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm

```diff
@@ -2125,24 +2147,41 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
         return ![webView _hasServiceWorkerForegroundActivityForTesting] && ![webView _hasServiceWorkerBackgroundActivityForTesting];
     }));
 
-    waitUntilEvaluatesToTrue([&] { return [webView _webProcessState] == _WKWebProcessStateSuspended; }, 500);
+    EXPECT_TRUE(waitUntilEvaluatesToTrue([&] { return [webView _webProcessState] == _WKWebProcessStateSuspended; }));
 
-    EXPECT_EQ([webView _webProcessState], _WKWebProcessStateSuspended);
+    // Due to the process assertion cache, the process can actually run for a little while after it
+    // enters the suspended state. Wait until the OS tells us the process is actually suspended.
+    pid_t pidBeforeTerminatingServiceWorker = [webView _webProcessIdentifier];
+    EXPECT_TRUE(waitUntilEvaluatesToTrue([&] { return isPIDSuspended(pidBeforeTerminatingServiceWorker); }));
 
     // Process with service worker and page is suspended, let's terminate the service worker.
-    auto typesToRemove = adoptNS([[NSSet alloc] initWithArray:@[WKWebsiteDataTypeServiceWorkerRegistrations]]);
-    [[WKWebsiteDataStore defaultDataStore] removeDataOfTypes:typesToRemove.get() modifiedSince:[NSDate distantPast] completionHandler:^() {
+    [[WKWebsiteDataStore defaultDataStore] removeDataOfTypes:[NSSet setWithObject:WKWebsiteDataTypeServiceWorkerRegistrations] modifiedSince:[NSDate distantPast] completionHandler:^() {
         done = true;
     }];
     TestWebKitAPI::Util::run(&done);
     done = false;
 
-    TestWebKitAPI::Util::runFor(1_s);
+    bool serviceWorkersAllTerminated = waitUntilEvaluatesToTrue([&]() {
+        __block bool done;
+        __block NSUInteger serviceWorkerCount;
+
+        [[WKWebsiteDataStore defaultDataStore] _runningOrTerminatingServiceWorkerCountForTesting:^(NSUInteger count) {
+            serviceWorkerCount = count;
+            done = true;
+        }];
+        TestWebKitAPI::Util::run(&done);
 
-    // Let's verify the WKWebView did not crash.
+        return !serviceWorkerCount;
+    }, 20);
+    EXPECT_TRUE(serviceWorkersAllTerminated);
```

Comment:
It would probably be nice to replace this with something like:

```
    EXPECT_TRUE(waitUntilEvaluatesToTrue([&] { return [processPool _serviceWorkerProcessCount] == 0; }));
```

But as far as I can tell, there is no connection between terminating a worker (even as a result of clearing the data store) and `_serviceWorkerProcessCount` decreasing. Is this expected?

Reply:
Yes, I think so. We remove the service worker process when there are no potential clients, see the call site of `removeContextConnectionIfPossible`